File: server/src/update.ts
import { Manifest, type VersionInfo } from "@ao3hub/shared";
import { rename, chmod, unlink } from "node:fs/promises";
import path from "node:path";
import { ARCH, BUILT_AT, PLATFORM, VERSION } from "./env";
import { loadConfig } from "./db";

function compareSemver(a: string, b: string): number {
  const pa = a.replace(/^v/, "").split(".").map((x) => parseInt(x, 10) || 0);
  const pb = b.replace(/^v/, "").split(".").map((x) => parseInt(x, 10) || 0);
  for (let i = 0; i < Math.max(pa.length, pb.length); i++) {
    const x = pa[i] ?? 0;
    const y = pb[i] ?? 0;
    if (x !== y) return x - y;
  }
  return 0;
}

function platformMatch(asset: { platform: string; arch: string }): boolean {
  return asset.platform === PLATFORM && asset.arch === ARCH;
}

export async function fetchManifest(): Promise<{ manifest: Manifest | null; error?: string }> {
  const cfg = await loadConfig();
  const url = cfg.update.manifestURL.trim();
  if (!url) return { manifest: null, error: "未配置 manifest URL" };
  try {
    const res = await fetch(url, { headers: { "cache-control": "no-cache" } });
    if (!res.ok) return { manifest: null, error: `manifest fetch failed: ${res.status}` };
    const json = await res.json();
    const parsed = Manifest.safeParse(json);
    if (!parsed.success) return { manifest: null, error: "manifest schema 校验失败" };
    return { manifest: parsed.data };
  } catch (e) {
    return { manifest: null, error: e instanceof Error ? e.message : String(e) };
  }
}

export async function versionInfo(): Promise<VersionInfo> {
  const base: VersionInfo = {
    current: VERSION,
    platform: PLATFORM,
    arch: ARCH,
    builtAt: BUILT_AT,
  };
  const { manifest } = await fetchManifest();
  if (!manifest) return base;
  const asset = manifest.assets.find(platformMatch);
  const hasUpdate = compareSemver(manifest.version, VERSION) > 0;
  return {
    ...base,
    latest: {
      version: manifest.version,
      notes: manifest.notes,
      publishedAt: manifest.publishedAt,
      hasUpdate,
      downloadUrl: asset?.url,
    },
  };
}

async function sha256Hex(file: string): Promise<string> {
  const buf = await Bun.file(file).arrayBuffer();
  const hash = await crypto.subtle.digest("SHA-256", buf);
  return Array.from(new Uint8Array(hash))
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

export type ApplyResult = {
  ok: boolean;
  version?: string;
  message: string;
  restart?: boolean;
};

export async function applyUpdate(opts: { force?: boolean } = {}): Promise<ApplyResult> {
  const { manifest, error } = await fetchManifest();
  if (!manifest) return { ok: false, message: error ?? "无法获取 manifest" };
  if (!opts.force && compareSemver(manifest.version, VERSION) <= 0) {
    return { ok: false, message: `当前 ${VERSION} 已是最新（remote ${manifest.version}）` };
  }
  const asset = manifest.assets.find(platformMatch);
  if (!asset) {
    return { ok: false, message: `manifest 中无 ${PLATFORM}/${ARCH} 资源` };
  }

  const exec = process.execPath;
  const dir = path.dirname(exec);
  const tmp = path.join(dir, `.ao3-hub.new-${Date.now()}`);

  const res = await fetch(asset.url);
  if (!res.ok) return { ok: false, message: `下载失败: ${res.status}` };
  const buf = await res.arrayBuffer();
  await Bun.write(tmp, buf);

  if (asset.sha256) {
    const sum = await sha256Hex(tmp);
    if (sum.toLowerCase() !== asset.sha256.toLowerCase()) {
      await unlink(tmp).catch(() => {});
      return { ok: false, message: `sha256 校验失败 expected=${asset.sha256} got=${sum}` };
    }
  }

  await chmod(tmp, 0o755);

  const backup = path.join(dir, `.ao3-hub.bak-${VERSION}`);
  try {
    await rename(exec, backup);
  } catch {}
  await rename(tmp, exec);

  return {
    ok: true,
    version: manifest.version,
    message: `已升级到 ${manifest.version}，进程即将退出，等待 launcher 重启`,
    restart: true,
  };
}

export function scheduleExit(delayMs = 600): void {
  setTimeout(() => process.exit(0), delayMs);
}

File: internal/app/update.go
package app

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"
)

func compareSemver(a, b string) int {
	pa := strings.Split(strings.TrimPrefix(a, "v"), ".")
	pb := strings.Split(strings.TrimPrefix(b, "v"), ".")
	n := len(pa)
	if len(pb) > n {
		n = len(pb)
	}
	for i := 0; i < n; i++ {
		x, y := 0, 0
		if i < len(pa) {
			x, _ = strconv.Atoi(pa[i])
		}
		if i < len(pb) {
			y, _ = strconv.Atoi(pb[i])
		}
		if x != y {
			return x - y
		}
	}
	return 0
}

func isDevVersion(version string) bool {
	return strings.HasPrefix(version, "dev-")
}

func compareVersions(a, b string) int {
	if a == b {
		return 0
	}
	if isDevVersion(a) || isDevVersion(b) {
		return 1
	}
	return compareSemver(a, b)
}

func platformMatch(asset ManifestAsset) bool {
	return asset.Platform == platformName() && asset.Arch == archName()
}

func manifestURLForChannel(channel string) string {
	if strings.ToLower(strings.TrimSpace(channel)) == "dev" {
		return DefaultDevUpdateManifestURL
	}
	return DefaultUpdateManifestURL
}

func resolveManifestURL(cfg Config) string {
	url := strings.TrimSpace(cfg.Update.ManifestURL)
	if url == "" || url == DefaultUpdateManifestURL || url == DefaultDevUpdateManifestURL {
		return manifestURLForChannel(cfg.Update.Channel)
	}
	return url
}

func (a *App) FetchManifest() (*Manifest, string) {
	cfg, err := a.store.LoadConfig()
	if err != nil {
		return nil, err.Error()
	}
	url := resolveManifestURL(cfg)
	if url == "" {
		return nil, "未配置 manifest URL"
	}
	req, err := http.NewRequest(http.MethodGet, url, nil)
	if err != nil {
		return nil, err.Error()
	}
	req.Header.Set("cache-control", "no-cache")
	res, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, err.Error()
	}
	defer res.Body.Close()
	if res.StatusCode < 200 || res.StatusCode >= 300 {
		return nil, fmt.Sprintf("manifest fetch failed: %d", res.StatusCode)
	}
	var manifest Manifest
	if err := json.NewDecoder(res.Body).Decode(&manifest); err != nil {
		return nil, "manifest schema 校验失败"
	}
	if manifest.Version == "" || manifest.Assets == nil {
		return nil, "manifest schema 校验失败"
	}
	return &manifest, ""
}

func (a *App) VersionInfo() VersionInfo {
	base := VersionInfo{
		Current:  Version,
		Platform: platformName(),
		Arch:     archName(),
		BuiltAt:  BuiltAt,
	}
	manifest, _ := a.FetchManifest()
	if manifest == nil {
		return base
	}
	var asset *ManifestAsset
	for i := range manifest.Assets {
		if platformMatch(manifest.Assets[i]) {
			asset = &manifest.Assets[i]
			break
		}
	}
	latest := &LatestVersion{
		Version:     manifest.Version,
		Channel:     manifest.Channel,
		Notes:       manifest.Notes,
		PublishedAt: manifest.PublishedAt,
		HasUpdate:   compareVersions(manifest.Version, Version) > 0,
	}
	if asset != nil {
		latest.DownloadURL = asset.URL
	}
	base.Latest = latest
	return base
}

func sha256HexFile(file string) (string, error) {
	f, err := os.Open(file)
	if err != nil {
		return "", err
	}
	defer f.Close()
	h := sha256.New()
	if _, err := io.Copy(h, f); err != nil {
		return "", err
	}
	return hex.EncodeToString(h.Sum(nil)), nil
}

type ApplyResult struct {
	OK      bool   `json:"ok"`
	Version string `json:"version,omitempty"`
	Message string `json:"message"`
	Restart bool   `json:"restart,omitempty"`
}

func (a *App) ApplyUpdate(force bool) ApplyResult {
	manifest, fetchErr := a.FetchManifest()
	if manifest == nil {
		if fetchErr == "" {
			fetchErr = "无法获取 manifest"
		}
		return ApplyResult{OK: false, Message: fetchErr}
	}
	if !force && compareVersions(manifest.Version, Version) <= 0 {
		return ApplyResult{OK: false, Message: fmt.Sprintf("当前 %s 已是最新（remote %s）", Version, manifest.Version)}
	}
	var asset *ManifestAsset
	for i := range manifest.Assets {
		if platformMatch(manifest.Assets[i]) {
			asset = &manifest.Assets[i]
			break
		}
	}
	if asset == nil {
		return ApplyResult{OK: false, Message: fmt.Sprintf("manifest 中无 %s/%s 资源", platformName(), archName())}
	}
	exec, err := os.Executable()
	if err != nil {
		return ApplyResult{OK: false, Message: err.Error()}
	}
	dir := filepath.Dir(exec)
	tmp := filepath.Join(dir, fmt.Sprintf(".ao3-hub.new-%d", time.Now().UnixMilli()))

	res, err := http.Get(asset.URL)
	if err != nil {
		return ApplyResult{OK: false, Message: err.Error()}
	}
	defer res.Body.Close()
	if res.StatusCode < 200 || res.StatusCode >= 300 {
		return ApplyResult{OK: false, Message: fmt.Sprintf("下载失败: %d", res.StatusCode)}
	}
	out, err := os.OpenFile(tmp, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0o755)
	if err != nil {
		return ApplyResult{OK: false, Message: err.Error()}
	}
	if _, err := io.Copy(out, res.Body); err != nil {
		_ = out.Close()
		_ = os.Remove(tmp)
		return ApplyResult{OK: false, Message: err.Error()}
	}
	_ = out.Close()

	if asset.SHA256 != "" {
		sum, err := sha256HexFile(tmp)
		if err != nil {
			_ = os.Remove(tmp)
			return ApplyResult{OK: false, Message: err.Error()}
		}
		if !strings.EqualFold(sum, asset.SHA256) {
			_ = os.Remove(tmp)
			return ApplyResult{OK: false, Message: fmt.Sprintf("sha256 校验失败 expected=%s got=%s", asset.SHA256, sum)}
		}
	}
	_ = os.Chmod(tmp, 0o755)
	backup := filepath.Join(dir, ".ao3-hub.bak-"+Version)
	_ = os.Rename(exec, backup)
	if err := os.Rename(tmp, exec); err != nil {
		return ApplyResult{OK: false, Message: err.Error()}
	}
	return ApplyResult{
		OK:      true,
		Version: manifest.Version,
		Message: fmt.Sprintf("已升级到 %s，进程即将退出，等待 launcher 重启", manifest.Version),
		Restart: true,
	}
}

func scheduleExit(delayMS int) {
	if delayMS < 0 {
		delayMS = 600
	}
	go func() {
		time.Sleep(time.Duration(delayMS) * time.Millisecond)
		os.Exit(0)
	}()
}
